src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs

```diff
@@ -0,0 +1,271 @@
+    /// <returns>Returns a task that completes when the document is successfully saved.</returns>
+    public async Task AddDocumentAsync(SearchableDocument document, CancellationToken cancellationToken)
+    {
+        if (document.ConnectionInfo is null)
+        {
+            throw new ArgumentNullException(nameof(document.ConnectionInfo));
+        }
+        string connectionString = this.configuration.GetValue<string>(document.ConnectionInfo.ConnectionName);
+
+        MongoClient searchIndexClient = await this.GetMongoClientAsync(connectionString);
+
+        this.CreateVectorIndexIfNotExists(searchIndexClient, document.ConnectionInfo.CollectionName ?? defaultSearchIndexName);
+
+        await this.UpsertVectorAsync(searchIndexClient, document, document.ConnectionInfo.CollectionName ?? defaultSearchIndexName);
+    }
+
+    /// <summary>
+    /// Search for documents using the provided request.
+    /// </summary>
+    /// <param name="request">The search request.</param>
+    /// <returns>Search Response.</returns>
+    /// <exception cref="ArgumentException">Throws argument exception if query or embeddings is null.</exception>
+    /// <exception cref="InvalidOperationException">Throws the invalid operation exception if search result response is null.</exception>
+    public async Task<SearchResponse> SearchAsync(SearchRequest request)
+    {
+        if (request.Query is null && request.Embeddings.IsEmpty)
+        {
+            throw new ArgumentException("Either query or embeddings must be provided");
+        }
+
+        if (request.ConnectionInfo is null)
+        {
+            throw new ArgumentNullException(nameof(request.ConnectionInfo));
+        }
+
+        string connectionString = this.configuration.GetValue<string>(request.ConnectionInfo.ConnectionName);
+        MongoClient searchIndexClient = await this.GetMongoClientAsync(connectionString);
+
+        try
+        {
+            IMongoCollection<BsonDocument> collection = searchIndexClient.GetDatabase(databaseName).GetCollection<BsonDocument>(request.ConnectionInfo.CollectionName ?? defaultSearchIndexName);
+
+            // Search Azure Cosmos DB for MongoDB vCore collection for similar embeddings and project fields.
+            BsonDocument[] pipeline = new BsonDocument[]
+            {
+                new BsonDocument
+                {
+                    {
+                        "$search", new BsonDocument
+                        {
+                            {
+                                "cosmosSearch", new BsonDocument
+                                {
+                                    { "vector", !(request.Embeddings.IsEmpty) ? new BsonArray(request.Embeddings.ToArray()) : new BsonArray()},
+                                    { "path", "embedding" },
+                                    { "k", this.isSemanticSearchEnabled && !request.Embeddings.IsEmpty ? Math.Max(50, request.MaxResults) : request.MaxResults }
```

Comment:
does Cosmos have similar capabilities as AI Search for semantic ranking? https://learn.microsoft.com/en-us/azure/search/semantic-search-overview

We shouldn't be using this if its not a special feature of Cosmos and default to just MaxResults

Reply:
no it doesn't let me remove it